From b8b98ba4b5ee61e04ed5f64fd3f6fe6ee2f369e9 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:13:39 -0400
Subject: [powerpc] module: Fix TOC symbol CRC

Message-id: <1410545655-205645-591-git-send-email-dzickus@redhat.com>
Patchwork-id: 94650
O-Subject: [RHEL7 PATCH 590/626] powerpc/module: Fix TOC symbol CRC
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit c2cbcf533a1d7443cf1d6aae1127491792601587
Author: Laurent Dufour <ldufour@linux.vnet.ibm.com>
Date:   Tue Jun 24 10:53:59 2014 +0200

    powerpc/module: Fix TOC symbol CRC

    The commit 71ec7c55ed91 introduced the magic symbol ".TOC." for ELFv2 ABI.
    This symbol is built manually and has no CRC value computed. A zero value
    is put in the CRC section to avoid modpost complaining about a missing CRC.
    Unfortunately, this breaks the kernel module loading when the kernel is
    relocated (kdump case for instance) because of the relocation applied to
    the kcrctab values.

    This patch compute a CRC value for the TOC symbol which will match the one
    compute by the kernel when it is relocated - aka '0 - relocate_start' done in
    maybe_relocated called by check_version (module.c).

    Signed-off-by: Laurent Dufour <ldufour@linux.vnet.ibm.com>
    Cc: Anton Blanchard <anton@samba.org>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/module_64.c b/arch/powerpc/kernel/module_64.c
index 077d2ce..d807ee6 100644
--- a/arch/powerpc/kernel/module_64.c
+++ b/arch/powerpc/kernel/module_64.c
@@ -315,8 +315,17 @@ static void dedotify_versions(struct modversion_info *vers,
  struct modversion_info *end;
 
  for (end = (void *)vers + size; vers < end; vers++)
-  if (vers->name[0] == '.')
+  if (vers->name[0] == '.') {
    memmove(vers->name, vers->name+1, strlen(vers->name));
+#ifdef ARCH_RELOCATES_KCRCTAB
+   /* The TOC symbol has no CRC computed. To avoid CRC
+    * check failing, we must force it to the expected
+    * value (see CRC check in module.c).
+    */
+   if (!strcmp(vers->name, "TOC."))
+    vers->crc = -(unsigned long)reloc_start;
+#endif
+  }
 }
 
 /* Undefined symbols which refer to .funcname, hack to funcname (or .TOC.) */
-- 
1.7.1